perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java

```diff
@@ -1091,66 +1086,54 @@ private void setFromAndReplyTo(MimeMessage message, String fromName, String from
-	 * In case no such admins not set, use VO attribute.
-	 * If Vo attribute not set (both group or vo application), notify all corresponding vo admins (admin roles configured to receive VO notifications).
-	 * Otherwise, BACKUP_FROM address will be used.
+	 * If group attribute is not set and Application is a group application, notify all corresponding group admins
+	 * (admin roles configured to receive GROUP notifications).
+	 * In case no such admins are set, try the same procedure on parent group (if exists)
+	 * When reaches the top-level group and no recipients have been identified, repeat the same procedure on VO level (without cascading).
+	 * When no recipient is identified on the VO level or some error happens, BACKUP_FROM address will be used.
 	 *
 	 * @param app application to decide if it's VO or Group application
 	 * @return list of mail addresses to send mail to
 	 */
-	private List<String> getToMailAddresses(Application app) {
-		List<String> result = new ArrayList<>();
-
-		// get proper value from attribute
+	private Set<String> getToMailAddresses(Application app) {
+		Set<String> result = new HashSet<>();
 		try {
-			Attribute attrSenderEmail;
-
 			if (app.getGroup() == null) {
 				// it is a VO application
-				return getMailsFromVo(app, URN_VO_TO_EMAIL);
-			}
-
-			attrSenderEmail = attrManager.getAttribute(registrarSession, app.getGroup(), URN_GROUP_TO_EMAIL);
-
-			if (attrSenderEmail == null || attrSenderEmail.getValue() == null) {
-				// not specified toEmail group attribute
-
-				// try to use all group admins with specified preferred emails
-				List<String> rolesToNotify = getNotificationReceiverRoles(GROUP);
-				List<RichUser> admins = new ArrayList<>();
-				for (String role : rolesToNotify) {
-					admins.addAll(AuthzResolverBlImpl.getRichAdmins(registrarSession, app.getGroup(), role));
-				}
-				result.addAll(getUserPreferredMails(registrarSession, admins));
-				result = result.stream().distinct().toList();
-
-				if (!result.isEmpty()) {
-					return result;
-				}
-				return getMailsFromVo(app, URN_VO_TO_EMAIL);
+				return getToEmailsVoLevel(app.getVo());
 			}
-
-			// specified toEmail group attribute, use valid emails
-			ArrayList<String> value = attrSenderEmail.valueAsList();
-			for (String adr : value) {
-				if (adr != null && !adr.isEmpty()) {
-					result.add(adr);
+			// we have group application
+			Group currentGroup = app.getGroup();
+			do {
+				result = getToEmailsGroupLevel(currentGroup);
+				if (result.isEmpty()) {
+					try {
+						currentGroup = groupsManager.getGroupById(registrarSession, currentGroup.getParentGroupId());
+					} catch (GroupNotExistsException e) {
+						log.error("[MAIL MANAGER] Inconsistency detected - parent (in hierarchy) group with ID '{}' of group '{}' does not exist",
+							currentGroup.getParentGroupId(), app.getGroup());
+						break;
+					}
```

Comment:
I think this does not have to be inconsistency, since there is no check if the group has a parent.

Reply:
I have added a check for null before the call. What I meant by the log is that although `currentGroup` has got `parentGroupId`, a group for such ID could not be found, which IMHO means there is some inconsistency.

Reply:
Yes with the check in place this does make sense. However I believe that now if the result is not empty for first group we get an infinite loop. Maybe we could change the condition to `!result.isEmpty()` since the current one can never actually end the loop. It only breaks through the exception.

Reply:
You are right, well spotted :) will add a the clause into the loop condition.